PPS lhcInfo split: per fill and per LS PopCon

JanChyczynski opened the pull request:
#### PR description:

Populator of Conditions for the new records, `LHCInfoPerLS` and `LHCInfoPerFill` introduced in [PR39495](https://github.com/cms-sw/cmssw/pull/39495), to be used instead of `LHCInfo`. Manages the data sources (OMS, CTPPS OnlineDB, DIP), creates objects of the new classes and writes them to the ofline database (CondDB). Heavily based on original `LHCInfo PopCon`.


The main mechanism is to take fills data from `OMS` (filtered by time and having stable beam), iterate through them and for every fill (meeting certian coditions depending on the mode and record) get data from all the sources, create payloads and write them to the offline database.

There are separate programs for `...perLS` and `...perFill` classess. Both of them have 2 modes: `duringFill` and `endFill`.

The `duringFill` mode, when ran, writes data (just 1 payload of most recent data) only for ongoing fills and only during stable beam. `PerFill` PopCon writes the payload only once per fill, and `perLS` PopCon writes a payload every time it's run during a stable beam.

The `endFill` processes a fill only once, after it's ended. `PerFill` PopCon writes payloads corresponding to all the lumisections of the fill (filtered for duplicates), and `perLS` PopCon writes 2 payloads: corresponding to the start, and the end of the stable beam of the fill.

The worklows are ment to be run by python scripts located in `CondTools/RunInfo/python`:
```
LHCInfoPerFillPopConAnalyzerDuringFill.py
LHCInfoPerFillPopConAnalyzerEndFill.py
LHCInfoPerLSPopConAnalyzerDuringFill.py
LHCInfoPerLSPopConAnalyzerEndFill.py
```
more info: [presentation on LHCInfo update from AlCaDBmeeting](https://indico.cern.ch/event/1200061/contributions/5046070/attachments/2524882/4342370/LHCInfo%20final%20pres%20for%20AlCa.pdf)

#### PR validation:

The python scripts running the workflow were tested on several defferent periods of time (set by the `startTime` and `endTime` parameters) by analysing the logs and data in the written payloads.

```
cd CondTools/RunInfo/python
cmsRun LHCInfoPerFillPopConAnalyzerDuringFill.py = tag=lhcInfoPerFill_during
cmsRun LHCInfoPerFillPopConAnalyzerEndFill.py = tag=lhcInfoPerFill_end
cmsRun LHCInfoPerLSPopConAnalyzerDuringFill.py = tag=lhcInfoPerLS_during
cmsRun LHCInfoPerLSPopConAnalyzerEndFill.py = tag=lhcInfoPerLS_end
```

To run the PopCon the PPS database keys are required (located in `.cms_cond` folder in path you can specify in `authenticationPath` argument in the python script).

The new PopCon was also used in further development (update of proton reconstruction, to be submitted in the next PR).

JanChyczynski committed:
lhcInfo split: per fill and per LS PopCon

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/34268

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h  modified in PR(s): #39685
  - File  CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h  modified in PR(s): #39685
  - File  CondTools/RunInfo/interface/OMSAccess.h  modified in PR(s): #39685
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #39685
  - File  CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc  modified in PR(s): #39685
  - File  CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc  modified in PR(s): #39685
  - File  CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzerEndFill.py  modified in PR(s): #39685
  - File  CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzerEndFill.py  modified in PR(s): #39685
  - File  CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc  modified in PR(s): #39685
  - File  CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc  modified in PR(s): #39685
  - File  CondTools/RunInfo/src/OMSAccess.cc  modified in PR(s): #39685

cmsbuild commented:
A new Pull Request was created by  @JanChyczynski (jan_chyczynski) for master.

It involves the following packages:

- CondTools/RunInfo (**db**)


@malbouis, @cmsbuild, @saumyaphor4252, @ggovi, @francescobrivio, @tvami can you please review it and eventually sign? Thanks.
@mmusich this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

tvami commented on CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
```diff
@@ -0,0 +1,770 @@
+}
+//L1: try with different m_dipSchema
+//L2: try with different m_name
+LHCInfoPerFillPopConSourceHandler::~LHCInfoPerFillPopConSourceHandler() {}
```
```suggestion
LHCInfoPerFillPopConSourceHandler::~LHCInfoPerFillPopConSourceHandler() = default;
```

tvami commented on CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
```diff
@@ -0,0 +1,770 @@
+//L2: try with different m_name
+LHCInfoPerFillPopConSourceHandler::~LHCInfoPerFillPopConSourceHandler() {}
+
+namespace LHCInfoPerFillImpl {
```
```suggestion
namespace theLHCInfoPerFillImpl {
```
rule 2.7 https://cms-sw.github.io/cms_coding_rules.html

tvami commented on CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h:
```diff
@@ -0,0 +1,65 @@
+#ifndef LHCINFOPOPCONSOURCEHANDLER_H
+#define LHCINFOPOPCONSOURCEHANDLER_H
```
same header comment

forthommel replied:
Hi @tvami,
Never considered adding `SortIncludes: true` to [.clang-format](https://github.com/cms-sw/cmssw/blob/master/.clang-format)?

tvami replied:
That's beyond me, @smuzaffar is there a reason for this set to `false` in default?

forthommel replied:
Of course, that would require a _lot_ of 
```cpp
// clang-format off
...
// clang-format on
```
everywhere...

smuzaffar replied:
I do not remember the exact reasoning behind explicitly setting `SortIncludes: false` (may be @davidlange6 remembers? ) but my quess is this was done to avoid re-formatting all of cmssw files.

davidlange6 replied:
it is a change that can reveal code bugs.. Likely no one had the ambition to fix all the bugs that code format change would uncover. But I guess its a change many would like to have done..

tvami commented on CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h:
```diff
@@ -0,0 +1,62 @@
+
+#include <string>
+
+#include "CondCore/PopCon/interface/PopConSourceHandler.h"
+#include "CondFormats/RunInfo/interface/LHCInfoPerFill.h"
+#include "FWCore/ParameterSet/interface/ParameterSetfwd.h"
+#include "CondTools/RunInfo/interface/OMSAccess.h"
```
please alpha order this

tvami commented on CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h:
```diff
@@ -0,0 +1,65 @@
+
+#include <string>
+
+#include "CondCore/PopCon/interface/PopConSourceHandler.h"
+#include "CondFormats/RunInfo/interface/LHCInfoPerLS.h"
+#include "FWCore/ParameterSet/interface/ParameterSetfwd.h"
+#include "CondCore/CondDB/interface/Types.h"
+#include "CondTools/RunInfo/interface/OMSAccess.h"
```
please alpha order this

tvami commented on CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc:
```diff
@@ -0,0 +1,535 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
```
alpha order

tvami commented on CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc:
```diff
@@ -0,0 +1,535 @@
+}
+//L1: try with different m_dipSchema
+//L2: try with different m_name
+LHCInfoPerLSPopConSourceHandler::~LHCInfoPerLSPopConSourceHandler() {}
```
default as earlier

tvami commented on CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
```diff
@@ -0,0 +1,770 @@
+  }
+  boost::posix_time::ptime now = boost::posix_time::second_clock::local_time();
+  m_endTime = now;
+  if (pset.exists("endTime")) {
```
exists checks on the PSet are discouraged, please do a `fillDescription` instead, see more in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

JanChyczynski replied:
I've written `fillDescritption` in `LHCInfoPerLS(Fill)PopConSourceHandler` according to the twiki page but later I realised it doesn't work. The `_cfi.py` file is not created during `scram b` nor using `edmWriteConfigs`. When running the new PopCon the parameters aren't getting checked and the default values are not passed (in practice the former optional parameters bacame obligatory).

I think it's because LHCInfoPerLSPopConSourceHandler and LHCInfoPerFillPopConSourceHandler are not plugins by themselves and the parameters are not passed to them directly. They are used to instantiate a `PopConAnalyzer` template in `LHCInfoPerLS(Fill)PopConAnalyzer.cc`:
```
typedef popcon::PopConAnalyzer<LHCInfoPerLSPopConSourceHandler> LHCInfoPerLSPopConAnalyzer;
```
and the parameters passing is implemented in https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/CondCore/PopCon/interface/PopConAnalyzer.h#L22
So the `analyzer` receives, among others, nested `ParameterSet` parameter called "Source" and passes it to the `sourceHandler` and the sourceHandler's `fillDescritption` is ignored. 

I think that for the `fillDescritpion` to work in that case it would have to be written for the `LHCInfoPerLS(Fill)PopConAnalyzer` but I see a couple of difficulties with that:
- defining the `LHCInfoPerLS(Fill)PopConAnalyzer` couldn't be a simple typedef anymore. Might it create some problems?
- writing the `fillDescription` for the `LHCInfoPerLS(Fill)PopConAnalyzer` we would have to precisely describe the parameters not only for the `sourceHandler` but for the rest that arent passed to the `sourceHandler` too. I don't know what should be the requirements for these parameters so I would need some help with that.

Am I correct about this? What should we do: write the `fillDescription` for the analyzers or just go back to not using `fillDescription` in this case? From what I've seen there are no `fillDescriptions` to other PopCons.

BTW: using `fillDescription` I still have to make an `exists` check on the PSet but I changed it into `existsAs<std::string>` as suggested in the twiki page ("optional parameters" paragraph)

tvami replied:
Ahh yeah this is a complication :/ like this one needs the `fillPSetDescription` instead of the `fillDescription` as explained in the "A Plugin Module Using Another Helper Plugin" part of the twiki: 
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He

JanChyczynski replied:
Ok, I'll try to do it this way.

Regarding the `existsAs` in itself: I only changed it to `exsitsAs` from `exists` it as it was the suggested way in the twiki page:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Optional_Parameters

I don't see how it's affected by having the fillDesc but I could remove the `existsAs` by adding a default value (eg. empty string) and checking if the argument is empty instead of checking if it exists. Actually for `endTime` I'm checking both cases because I needed the empty string to work like this to be able to have it as an optional parameters of the python scripts that are running the PopCon.

tvami replied:
hi @JanChyczynski do you have an update to this?

tvami replied:
@makortel can you please clarify the points above?

makortel replied:
So apparently the instantiations of `popcon::PopConAnalyzer<S>` class template are the actual framework modules, and the `PopConAnalyzer` has a member variable of type `S` directly. Therefore this case is **not** "a plugin module using another helper plugin".

From this it also follows (as already mentioned in https://github.com/cms-sw/cmssw/pull/40817#discussion_r1157825190) that the current `LHCInfoPerFillPopConSourceHandler::fillDescriptions()` and `LHCInfoPerLSPopConSourceHandler::fillDescriptions()` do not get called.

What should be done instead is to
* define the `fillDescriptions(edm::ConfigurationDescriptions& descriptions)` functions in `popcon::PopConAnalyzer<S>`
* define `fillPSetDescription(edm::ParameterSetDescription& desc)` in `LHCInfoPerFillPopConSourceHandler` and `LHCInfoPerLSPopConSourceHandler`
* make `popcon::PopConAnalyzer<S>::fillDescriptions()` call `S::fillPSetDescription()`
  * it follows that either all `S` must have `fillPSetDescription()` implemented, or one adds code in `popcon::PopConAnalyzer<S>::fillDescriptions()` to have different behavior based on `S` having `fillPSetDescription()` or not

(I'd argue this would be complicated enough to warrant its own PR)

tvami replied:
Thanks Matti! Sure let's do this in another PR then, I can make a reminder issue about it

JanChyczynski replied:
Ok, I'll revert the changes with fillDescriptions for now then. 

So, should I remove the `existsAs` and replace it with just checking for a default value? I don't see how it's dependent on the fillDescriptions and don't really se how that's better then the `existsAs` check.

makortel replied:
Generally the experience has been that the application behavior is easier to understand when the behavior depends only on values of configuration parameters rather than also from the existence of parameters. E.g. `edmConfigDump` lists all parameters (that may describe the module's behavior enough for not having to look the code), overriding parameter values in python code can be done without explicit types on the right hand side (allows catching typos earlier), etc.

Written that, there are some legitimate cases for `exists` (or similar). To large degree it depends on the application (e.g. HLT does not work with optional parameters at all, RECO wants to avoid and asked for the static analyzer check).

tvami reviewed (commented):
Let me send you a partial review. I also didn't comment on each places, but there are a few recurring problems, please fix them at each occurance

vavati commented:
@tvami : Several data are coming from PPS database -> PopCon -> LHCInfo

smuzaffar commented:
bot do have `~/.cms_cond/db.key` but I do not know if it contains PPS keys or not

francescobrivio commented:
> bot has `~/.cms_cond/db.key` but I do not know if it contains PPS keys or not

ok I _think_ I have fixed this (i.e. given the permission to read the needed DBs to the IB key), so the next round of tests (after the comments are addressed) should work fine.

tvami commented:
> > bot has `~/.cms_cond/db.key` but I do not know if it contains PPS keys or not
> 
> ok I _think_ I have fixed this (i.e. given the permission to read the needed DBs to the IB key), so the next round of tests (after the comments are addressed) should work fine.

Great! So @JanChyczynski let's add unit tests that runs 
> LHCInfoPerFillPopConAnalyzerDuringFill.py
> LHCInfoPerFillPopConAnalyzerEndFill.py
> LHCInfoPerLSPopConAnalyzerDuringFill.py
> LHCInfoPerLSPopConAnalyzerEndFill.py

in CMSSW

tvami commented:
@JanChyczynski do you have any updates on this PR?

JanChyczynski commented:
> @JanChyczynski do you have any updates on this PR?

I made all the suggested changes but I'm still working on the unit tests.

tvami commented on CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
```diff
@@ -68,36 +70,54 @@ namespace cond {
       m_prevPayload(),
       m_tmpBuffer() {
-  if (pset.exists("startTime")) {
+  if (pset.existsAs<std::string>("startTime", false)) {
```
you dont need the `existsAs` since you have a fillDesc (and anyway `existsAs` is discouraged)

JanChyczynski commented:
I've just pushed a few commits with the suggested changes and the unit tests (I'll squash them after the review).
It will still need some work because of the problem with fillDescritpion I described in the answer to your comment.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35042

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #40817 was updated. @cmsbuild, @tvami, @saumyaphor4252, @francescobrivio can you please check and sign again.

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35044

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #40817 was updated. @cmsbuild, @tvami, @saumyaphor4252, @francescobrivio can you please check and sign again.

tvami commented:
@cmsbuild , please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b1fdd/31816/summary.html
**COMMIT**: 9da7a1cae8aec4a2bb2528ea68119d75e41fb5e2
**CMSSW**: CMSSW_13_1_X_2023-04-04-1600/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40817/31816/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test CondToolsLHCInfoNewPopConTest had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially added 10 lines to the logs
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3365163
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3365132
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 202 log files, 155 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-04-1600+1b1fdd/56411/triggerResults): no differences found

JanChyczynski commented:
I struggle to recreate and fix the failing unit test. It fails with `File "src/CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzerEndFill.py" could not be opened!` message and I don't understand why the file cannot be open beacause it exists and it's a relative path from the same place as the paths in the `test_LhcInfoTest.sh` (and the CondToolsLHCInfoTest passes).

On my "develop" enviroment CMSSW_13_0_0_pre4 both of these tests pass.
I cannot instantiate the enviroment the tests are run (CMSSW_13_1_X_2023-04-04-1600) because I get 
```ERROR: Unable to find release area for "CMSSW" version "CMSSW_13_1_X_2023-04-04-1600" for arch slc7_amd64_gcc10.
       Please make sure you have used the correct name/version.
       You can run "scram list CMSSW" to get the list of available versions.
```
(I tried to run it on `/eos/home-j/jchyczyn`)

I was able to instantiate CMSSW_13_1_X_2023-04-04-1100 where both of these tests are failing with similar error of `File "src/CondTools/RunInfo ...." could not be opened!`. It's weird as I don't see the CondToolsLHCInfoTest failing on the IB dashboard.

Any advice on fixing this test?

perrotta commented:
@smuzaffar could you please help @JanChyczynski with the problem he's encountering in the unit test? (see https://github.com/cms-sw/cmssw/pull/40817#issuecomment-1497752698)
The python file search must be adapted to the new styles...

smuzaffar commented:
@JanChyczynski , in CMSW_13_1_X all tests are now run in their own dedicated directory. So relative paths will not work. please use 
```
LOCAL_TEST_DIR=${CMSSW_BASE}/src/CondTools/RunInfo/test
SCRIPTS_DIR=${CMSSW_BASE}/src/CondTools/RunInfo/python
```

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35080

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #40817 was updated. @tvami, @saumyaphor4252, @francescobrivio can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b1fdd/31855/summary.html
**COMMIT**: 02783182a4987517c78a829c4d229dc47b127ffd
**CMSSW**: CMSSW_13_1_X_2023-04-05-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40817/31855/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 21 lines to the logs
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3365163
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3365132
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 202 log files, 155 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-05-1100+1b1fdd/56438/triggerResults): no differences found

makortel commented:
On a separate note, is there a real need to have the functionality of e.g. `LHCInfoPerFillPopConSourceHandler` spread in 3 files
* `CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h`
* `CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc`
* `CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc`
?

If the `LHCInfoPerFillPopConSourceHandler` class is used only to define the `LHCInfoPerFillPopConAnalyzer` plugin, it would be better to have all the code in one file `CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc`.

JanChyczynski commented:
> On a separate note, is there a real need to have the functionality of e.g. `LHCInfoPerFillPopConSourceHandler` spread in 3 files
> 
> * `CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h`
> * `CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc`
> * `CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc`
> 
> ?
> 
> If the `LHCInfoPerFillPopConSourceHandler` class is used only to define the `LHCInfoPerFillPopConAnalyzer` plugin, it would be better to have all the code in one file `CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc`.


I don't know if there's a good reason and I see how it can be confusing. On the other hand, we would loose that header-implementation structure (I don't know if it matters that much though). Also, all the PopConAnalyzers are implemented like this, in 3 files, including some unrelated to this PR (eg. `RunInfoPopConAnalyzer.cc`, `RunSummaryPopConAnalyzer.cc`). In order not to introduce inconsistencies I think we should change all of them, but is it still in the scope of this PR? (There are 7 PopConAnalyzers in total.)

makortel commented:
> I don't know if there's a good reason and I see how it can be confusing. On the other hand, we would loose that header-implementation structure (I don't know if it matters that much though).

The header-source structure would matter only if there would be many source files including the header. Otherwise our recommendation is to put all code into a single source file (6.7 in https://cms-sw.github.io/cms_coding_rules.html#6--packaging-rules-1).

> Also, all the PopConAnalyzers are implemented like this, in 3 files, including some unrelated to this PR (eg. `RunInfoPopConAnalyzer.cc`, `RunSummaryPopConAnalyzer.cc`). In order not to introduce inconsistencies I think we should change all of them, but is it still in the scope of this PR? (There are 7 PopConAnalyzers in total.)

I personally don't mind short-term inconsistencies, and would say to not extend the scope of this PR. But in the end it is up to @cms-sw/db-l2.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35236

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #40817 was updated. @cmsbuild, @tvami, @saumyaphor4252, @francescobrivio can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b1fdd/32055/summary.html
**COMMIT**: a48dafce6a573f86628b2091e9dd141d788ac86d
**CMSSW**: CMSSW_13_1_X_2023-04-19-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40817/32055/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 2 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3459877
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3459852
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-19-1100+1b1fdd/56594/triggerResults): no differences found

tvami commented:
hi @JanChyczynski please also squash the commits, thanks

JanChyczynski commented:
When merging the PopCon Analyzers into single files I wanted to extract some common code along the way. I wanted to move the implementation of  `LumiSectionFilter` into `CondTools/RunInfo/plugins/LumiSectionFilter.h` and use it in both new PopConAnalyzers but I got a linker error:
` ERROR:Private Header: File src/CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc via it's public headers in src/CondTools/RunInfo/interface includes src/CondTools/RunInfo/plugins/LumiSectionFilter.h`

I put this attempt to extract this on https://github.com/CTPPS/cmssw/tree/jc_lhcInfo_split_PopCon_lumiSectionFilter
 
How can I fix it?

tvami commented:
6.3 says
> Only header files that expose a public interface should go into the Package/Subpackage/interface directory.

Seems to be true for `LumiSectionFilter`, right?

makortel commented:
> > Only header files that expose a public interface should go into the Package/Subpackage/interface directory.
> 
> Seems to be true for `LumiSectionFilter`, right?

As far as I can tell from https://github.com/CTPPS/cmssw/tree/jc_lhcInfo_split_PopCon_lumiSectionFilter `LumiSectionFilter.h` is not part of the public interface of `CondTools/RunInfo` as it is used only in one source file in `CondTools/RunInfo/plugins`. I'm a bit puzzled why scram complains (the error is actually not from the linker, but from a check done by scram), as the organization follows 6.3, 6.4, and 6.7. @smuzaffar Do you have an idea why scram complains in this case?

smuzaffar commented:
@JanChyczynski , try running building in a fresh area or run `scram build clean`. If you still get the `Private Header` error then please let me know how to reproduce it

JanChyczynski commented:
> @JanChyczynski , try building in a fresh area or run `scram build clean` first. If you still get the `Private Header` error then please let me know how to reproduce it

It helped, thank you

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35247

- This PR adds an extra 80KB to repository

- Found files with invalid states:
  - CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, 300ca9f60d2da3efa53895be39d57247a5827726, 0e7214106dc4f906cc1b5aa78100d1a8861a061b
    - Deleted: df2b88b57af91df577805aa61f1a77a7859b04ac
  - CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, 300ca9f60d2da3efa53895be39d57247a5827726, 0e7214106dc4f906cc1b5aa78100d1a8861a061b
    - Deleted: df2b88b57af91df577805aa61f1a77a7859b04ac
  - CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzerDuringFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e
  - CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzerEndFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e
  - CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzerDuringFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e
  - CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, a48dafce6a573f86628b2091e9dd141d788ac86d, 300ca9f60d2da3efa53895be39d57247a5827726
    - Deleted: 0e7214106dc4f906cc1b5aa78100d1a8861a061b
  - CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, 9da7a1cae8aec4a2bb2528ea68119d75e41fb5e2, a48dafce6a573f86628b2091e9dd141d788ac86d, 300ca9f60d2da3efa53895be39d57247a5827726
    - Deleted: 0e7214106dc4f906cc1b5aa78100d1a8861a061b
  - CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzerEndFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35247/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35247/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40817/35248

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, 300ca9f60d2da3efa53895be39d57247a5827726, 0e7214106dc4f906cc1b5aa78100d1a8861a061b
    - Deleted: df2b88b57af91df577805aa61f1a77a7859b04ac
  - CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, 300ca9f60d2da3efa53895be39d57247a5827726, 0e7214106dc4f906cc1b5aa78100d1a8861a061b
    - Deleted: df2b88b57af91df577805aa61f1a77a7859b04ac
  - CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzerDuringFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e
  - CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzerEndFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e
  - CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzerDuringFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e
  - CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, a48dafce6a573f86628b2091e9dd141d788ac86d, 300ca9f60d2da3efa53895be39d57247a5827726
    - Deleted: 0e7214106dc4f906cc1b5aa78100d1a8861a061b
  - CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: 39a5818fc8c6d1b2721828d789a2277cd1f0ca42, c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2, b6d9ddeb7e62ea2d4e5bd8caa9f0553b8cf7f3a7, 9da7a1cae8aec4a2bb2528ea68119d75e41fb5e2, a48dafce6a573f86628b2091e9dd141d788ac86d, 300ca9f60d2da3efa53895be39d57247a5827726
    - Deleted: 0e7214106dc4f906cc1b5aa78100d1a8861a061b
  - CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzerEndFill.py:
    - Added: 74b79031a825bfd8ffd40df3066a192cb225e7a9
    - Modified: c6e271b7f8fc34f5a669935e0b52a04820ff00b2, 7177ea2faf381157e1518d7697a5bba4746021c2
    - Deleted: e56611ce2c904d45c2f230b339e584425c49be9e

cmsbuild commented:
Pull request #40817 was updated. @cmsbuild, @tvami, @saumyaphor4252, @francescobrivio can you please check and sign again.

JanChyczynski commented:
I made all the suggested changes and also merged the python scripts running the new PopCon into 2 files instead of 4 (modes are now its arguments, not separate files).

> hi @JanChyczynski please also squash the commits, thanks

I'll do it when the review will be considered done, ok?

tvami commented:
> when the review will be considered done

what else is missing? I thought you addressed all comments

JanChyczynski commented:
> > when the review will be considered done
> 
> what else is missing? I thought you addressed all comments

Nothing is missing, I addressed all the comments. I just didn't want to squash it as you might have wanted to look at the last commits in separate to see how I addressed the last issues. If you have no further comments I guess it's done and I'll go ahead and squash them

tvami commented:
@cmsbuild , please test

cmsbuild commented:
Pull request #40817 was updated. @tvami, @saumyaphor4252, @francescobrivio can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b1fdd/32132/summary.html
**COMMIT**: 6b35a1247b8746a63ac84cac14abe1e9a554a44f
**CMSSW**: CMSSW_13_1_X_2023-04-25-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40817/32132/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 7 lines to the logs
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460685
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460657
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-25-1100+1b1fdd/56658/triggerResults): no differences found

tvami commented:
+db
 * tests pass, we agreed to have a few follow-up PR

rappoccio commented:
+1
